File: fxgl/src/main/java/com/almasb/fxgl/entity/GameWorld.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.entity;

import com.almasb.fxgl.annotation.Spawns;
import com.almasb.fxgl.app.FXGL;
import com.almasb.fxgl.core.collection.Array;
import com.almasb.fxgl.core.collection.ObjectMap;
import com.almasb.fxgl.core.collection.UnorderedArray;
import com.almasb.fxgl.core.logging.Logger;
import com.almasb.fxgl.core.math.FXGLMath;
import com.almasb.fxgl.core.reflect.ReflectionUtils;
import com.almasb.fxgl.entity.component.BoundingBoxComponent;
import com.almasb.fxgl.entity.component.IDComponent;
import com.almasb.fxgl.entity.component.IrremovableComponent;
import com.almasb.fxgl.entity.component.TimeComponent;
import com.almasb.fxgl.gameplay.Level;
import com.almasb.fxgl.parser.tiled.TiledMap;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.geometry.Point2D;
import javafx.geometry.Rectangle2D;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

/**
 * Represents pure logical state of the game.
 * Manages all entities and allows queries.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public final class GameWorld {

    private static Logger log = Logger.get("FXGL.GameWorld");
    
    private Array<Entity> updateList;

    /**
     * List of entities added to the update list in the next tick.
     */
    private Array<Entity> waitingList;

    /**
     * List of entities in the world.
     */
    private List<Entity> entities;

    private GameWorldQuery query;

    /**
     * Constructs the world with initial entity capacity = 32.
     */
    public GameWorld() {
        this(32);
    }

    /**
     * @param initialCapacity initial entity capacity
     */
    public GameWorld(int initialCapacity) {
        updateList = new Array<>(initialCapacity);
        waitingList = new UnorderedArray<>(initialCapacity);
        entities = new ArrayList<>(initialCapacity);

        query = new GameWorldQuery(entities);

        log.debug("Game world initialized");
    }

    /**
     * The entity will be added to update list in the next tick.
     *
     * @param entity the entity to add to world
     */
    public void addEntity(Entity entity) {
        if (entity.isActive())
            throw new IllegalArgumentException("Entity is already attached to world");

        waitingList.add(entity);
        entities.add(entity);

        add(entity);
    }

    public void addEntities(Entity... entitiesToAdd) {
        for (Entity e : entitiesToAdd) {
            addEntity(e);
        }
    }

    /**
     * Entity will be removed from world and parties notified of removal in the same frame.
     * The components and controls of entity will be removed in the next frame to avoid
     * concurrency issues.
     */
    public void removeEntity(Entity entity) {
        if (!entity.isActive()) {
            log.warning("Attempted to remove entity which is not active");
            return;
        }

        if (!canRemove(entity))
            return;

        if (entity.getWorld() != this)
            throw new IllegalArgumentException("Attempted to remove entity not attached to this world");

        entities.remove(entity);

        entity.markForRemoval();
        notifyEntityRemoved(entity);

        // we cannot clean entities here because this may have been called through a control
        // while entity is being updated
        // so, we clean the entity on next frame
    }

    public void removeEntities(Entity... entitiesToRemove) {
        for (Entity e : entitiesToRemove) {
            removeEntity(e);
        }
    }

    /**
     * Performs a single world update tick.
     *
     * @param tpf time per frame
     */
    private void update(double tpf) {
        updateList.addAll(waitingList);
        waitingList.clear();

        for (Iterator<Entity> it = updateList.iterator(); it.hasNext(); ) {
            Entity e = it.next();

            if (!e.isActive()) {
                // clean entities removed in the last frame
                e.clean();
                it.remove();
            } else {
                TimeComponent time = e.getComponent(TimeComponent.class);

                double tpfRatio = time == null ? 1.0 : time.getValue();

                e.update(tpf * tpfRatio);
            }
        }
    }

    /**
     * Removes all (including with IrremovableComponent) entities.
     * Does NOT clear state listeners.
     * Do NOT call this method manually.
     * It is called automatically by FXGL during initGame().
     */
    public void clear() {
        log.debug("Clearing game world");

        waitingList.clear();

        // we may still have some entities that have been removed but not yet cleaned
        for (Iterator<Entity> it = updateList.iterator(); it.hasNext(); ) {
            Entity e = it.next();

            if (!e.isActive()) {
                e.clean();
            }

            it.remove();
        }

        // entities list does not contain "not active" entities, so we do full clean
        for (Iterator<Entity> it = entities.iterator(); it.hasNext(); ) {
            Entity e = it.next();

            e.markForRemoval();
            notifyEntityRemoved(e);
            e.clean();

            it.remove();
        }
    }

    private void add(Entity entity) {
        entity.init(this);
        notifyEntityAdded(entity);
    }

    private boolean canRemove(Entity entity) {
        return !entity.hasComponent(IrremovableComponent.class);
    }

    private Array<EntityWorldListener> worldListeners = new Array<>();

    public void addWorldListener(EntityWorldListener listener) {
        worldListeners.add(listener);
    }

    public void removeWorldListener(EntityWorldListener listener) {
        worldListeners.removeValueByIdentity(listener);
    }

    private void notifyEntityAdded(Entity e) {
        for (int i = 0; i < worldListeners.size(); i++) {
            worldListeners.get(i).onEntityAdded(e);
        }
    }

    private void notifyEntityRemoved(Entity e) {
        for (int i = 0; i < worldListeners.size(); i++) {
            worldListeners.get(i).onEntityRemoved(e);
        }
    }

    /**
     * @return direct list of entities in the world (do NOT modify)
     */
    public List<Entity> getEntities() {
        return entities;
    }

    /**
     * @return shallow copy of the entities list (new list)
     */
    public List<Entity> getEntitiesCopy() {
        return new ArrayList<>(entities);
    }

    public void onUpdate(double tpf) {
        update(tpf);
    }

    private ObjectProperty<Entity> selectedEntity = new SimpleObjectProperty<>();

    /**
     * @return last selected (clicked on by mouse) entity
     */
    public Optional<Entity> getSelectedEntity() {
        return Optional.ofNullable(selectedEntity.get());
    }

    /**
     * @return selected entity property
     */
    public ObjectProperty<Entity> selectedEntityProperty() {
        return selectedEntity;
    }

    /**
     * Set level to given.
     * Resets the world.
     * Adds all level entities to the world.
     *
     * @param level the level
     */
    public void setLevel(Level level) {
        clearLevel();

        log.debug("Setting level: " + level);
        level.getEntities().forEach(this::addEntity);
    }

    /**
     * @param mapFileName name of the .json file
     */
    public void setLevelFromMap(String mapFileName) {
        setLevelFromMap(FXGL.getAssetLoader().loadJSON(mapFileName, TiledMap.class));
    }

    public void setLevelFromMap(TiledMap map) {
        clearLevel();

        log.debug("Setting level from map");

        map.getLayers()
                .stream()
                .filter(l -> l.getType().equals("tilelayer"))
                .forEach(l -> Entities.builder()
                        .viewFromTiles(map, l.getName(), RenderLayer.BACKGROUND)
                        .buildAndAttach(this));

        map.getLayers()
                .stream()
                .filter(l -> l.getType().equals("objectgroup"))
                .flatMap(l -> l.getObjects().stream())
                .forEach(obj -> spawn(obj.getType(), new SpawnData(obj)));
    }

    /**
     * Removes removable entities.
     */
    private void clearLevel() {
        log.debug("Clearing removable entities");

        waitingList.clear();

        // we may still have some entities that have been removed but not yet cleaned
        // but we do not want to remove Irremovables
        for (Iterator<Entity> it = updateList.iterator(); it.hasNext(); ) {
            Entity e = it.next();

            if (canRemove(e)) {
                if (!e.isActive()) {
                    // clean it here because "entities" list does not have "e"
                    e.clean();
                }

                it.remove();
            }
        }

        // entities list does not contain "not active" entities, so we do full clean
        // but we do not want to remove Irremovables
        for (Iterator<Entity> it = entities.iterator(); it.hasNext(); ) {
            Entity e = it.next();

            if (canRemove(e)) {
                e.markForRemoval();
                notifyEntityRemoved(e);
                e.clean();

                it.remove();
            }
        }
    }

    private EntityFactory entityFactory = null;

    private ObjectMap<String, EntitySpawner> entitySpawners = new ObjectMap<>();

    /**
     * @return entity factory or null if not set
     */
    @SuppressWarnings("unchecked")
    public <T extends EntityFactory> T getEntityFactory() {
        return (T) entityFactory;
    }

    /**
     * Set main entity factory to be used via {@link GameWorld#spawn(String, SpawnData)}.
     *
     * @param entityFactory factory for creating entities
     */
    public void setEntityFactory(EntityFactory entityFactory) {
        this.entityFactory = entityFactory;

        entitySpawners.clear();

        ReflectionUtils.findMethodsMapToFunctions(entityFactory, Spawns.class, EntitySpawner.class)
                .forEach((annotation, entitySpawner) -> entitySpawners.put(annotation.value(), entitySpawner));
    }

    public Entity spawn(String entityName) {
        return spawn(entityName, 0, 0);
    }

    /**
     * Creates an entity with given name at x, y using specified entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by {@link Spawns}
     * @param position spawn location
     * @return spawned entity
     */
    public Entity spawn(String entityName, Point2D position) {
        return spawn(entityName, position.getX(), position.getY());
    }

    /**
     * Creates an entity with given name at x, y using specified entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by {@link Spawns}
     * @param x x position
     * @param y y position
     * @return spawned entity
     */
    public Entity spawn(String entityName, double x, double y) {
        return spawn(entityName, new SpawnData(x, y));
    }

    /**
     * Creates an entity with given name and data using specified entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by {@link Spawns}
     * @param data spawn data, such as x, y and any extra info
     * @return spawned entity
     */
    public Entity spawn(String entityName, SpawnData data) {
        if (entityFactory == null)
            throw new IllegalStateException("EntityFactory was not set!");

        EntitySpawner spawner = entitySpawners.get(entityName);
        if (spawner == null)
            throw new IllegalArgumentException("EntityFactory does not have a method annotated @Spawns(" + entityName + ")");

        Entity entity = spawner.spawn(data);
        addEntity(entity);
        return entity;
    }

    /* QUERIES */

    public <T extends Entity> EntityGroup<T> getGroup(Enum<?>... types) {
        return new EntityGroup<T>(this, (List<? extends T>) getEntitiesByType(types), types);
    }

    /**
     * Useful for singleton type entities, e.g. Player.
     * 
     * @return first occurrence matching given type
     */
    public Optional<Entity> getSingleton(Enum<?> type) {
        return getSingleton(e -> e.isType(type));
    }

    /**
     * @return first occurrence matching given predicate
     */
    public Optional<Entity> getSingleton(Predicate<Entity> predicate) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            if (predicate.test(e)) {
                return Optional.of(e);
            }
        }

        return Optional.empty();
    }

    /**
     * @param type entity type
     * @return a random entity with given type
     */
    public Optional<Entity> getRandom(Enum<?> type) {
        return FXGLMath.random(getEntitiesByType(type));
    }

    public Optional<Entity> getRandom(Predicate<Entity> predicate) {
        return FXGLMath.random(getEntitiesFiltered(predicate));
    }

    /**
     * @param type component type
     * @return array of entities that have given component
     */
    public List<Entity> getEntitiesByComponent(Class<? extends Component> type) {
        return entities.stream()
                .filter(e -> e.hasComponent(type))
                .collect(Collectors.toList());
    }

    /**
     * Returns a list of entities which are filtered by
     * given predicate.
     * Warning: object allocation.
     *
     * @param predicate filter
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesFiltered(Predicate<Entity> predicate) {
        return query.getEntitiesFiltered(predicate);
    }

    /**
     * GC-friendly version of {@link #getEntitiesFiltered(Predicate)}.
     *
     * @param result the array to collect entities
     * @param predicate filter
     */
    public void getEntitiesFiltered(Array<Entity> result, Predicate<Entity> predicate) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            if (predicate.test(e)) {
                result.add(e);
            }
        }
    }

    /**
     * If called with no arguments, all entities are returned.
     * Warning: object allocation.
     *
     * @param types entity types
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesByType(Enum<?>... types) {
        return query.getEntitiesByType(types);
    }

    /**
     * GC-friendly version of {@link #getEntitiesByType(Enum[])}.
     *
     * @param result the array to collect entities
     * @param types entity types
     */
    public void getEntitiesByType(Array<Entity> result, Enum<?>... types) {
        if (types.length == 0) {
            for (int i = 0; i < entities.size(); i++) {
                Entity e = entities.get(i);
                result.add(e);
            }
            return;
        }

        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            if (isOfType(e, types)) {
                result.add(e);
            }
        }
    }

    private boolean isOfType(Entity e, Enum<?>... types) {
        for (Enum<?> type : types) {
            if (e.isType(type)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Returns a list of entities
     * which are partially or entirely
     * in the specified rectangular selection.
     * Warning: object allocation.
     *
     * @param selection Rectangle2D that describes the selection box
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesInRange(Rectangle2D selection) {
        return query.getEntitiesInRange(selection);
    }

    /**
     * GC-friendly version of {@link #getEntitiesInRange(Rectangle2D)}.
     *
     * @param result the array to collect entities
     * @param minX min x
     * @param minY min y
     * @param maxX max x
     * @param maxY max y
     */
    public void getEntitiesInRange(Array<Entity> result, double minX, double minY, double maxX, double maxY) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            if (e.getBoundingBoxComponent().isWithin(minX, minY, maxX, maxY)) {
                result.add(e);
            }
        }
    }

    /**
     * Returns a list of entities
     * which colliding with given entity.
     *
     * Note: CollidableComponent is not considered.
     *
     * @param entity the entity
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getCollidingEntities(Entity entity) {
        return query.getCollidingEntities(entity);
    }

    /**
     * GC-friendly version of {@link #getCollidingEntities(Entity)}.
     *
     * @param result the array to collect entities
     * @param entity the entity
     */
    public void getCollidingEntities(Array<Entity> result, Entity entity) {
        BoundingBoxComponent entityBBox = entity.getBoundingBoxComponent();

        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            if (e.getBoundingBoxComponent().isCollidingWith(entityBBox) && e != entity) {
                result.add(e);
            }
        }
    }

    /**
     * Returns a list of entities which have the given render layer index.
     *
     * @param layer render layer
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesByLayer(RenderLayer layer) {
        return query.getEntitiesByLayer(layer);
    }

    /**
     * GC-friendly version of {@link #getEntitiesByLayer(RenderLayer)}.
     *
     * @param result the array to collect entities
     * @param layer render layer
     */
    public void getEntitiesByLayer(Array<Entity> result, RenderLayer layer) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);

            if (e.getRenderLayer().index() == layer.index()) {
                result.add(e);
            }
        }
    }

    /**
     * Returns a list of entities at given position.
     * The position x and y must equal to entity's position x and y.
     *
     * @param position point in the world
     * @return entities at given point
     */
    public List<Entity> getEntitiesAt(Point2D position) {
        return query.getEntitiesAt(position);
    }

    /**
     * GC-friendly version of {@link #getEntitiesAt(Point2D)}.
     *
     * @param result the array to collect entities
     * @param position point in the world
     */
    public void getEntitiesAt(Array<Entity> result, Point2D position) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);

            if (e.getPosition().equals(position)) {
                result.add(e);
            }
        }
    }

    /**
     * Returns the closest entity to the given entity with given
     * filter. The given
     * entity itself is never returned.
     * <p>
     * If there no entities satisfying the requirement, {@link Optional#empty()}
     * is returned.
     * Warning: object allocation.
     *
     * @param entity selected entity
     * @param filter requirements
     * @return closest entity to selected entity with type
     */
    public Optional<Entity> getClosestEntity(Entity entity, Predicate<Entity> filter) {
        Array<Entity> array = new UnorderedArray<>(64);

        for (Entity e : entities) {
            if (filter.test(e) && e != entity) {
                array.add(e);
            }
        }

        if (array.size() == 0)
            return Optional.empty();

        array.sort(Comparator.comparingDouble(e -> e.distance(entity)));

        return Optional.of(array.get(0));
    }

    /**
     * Returns an entity whose IDComponent matches given name and id.
     * <p>
     * Returns {@link Optional#empty()} if no entity was found with such combination.
     * This query only works on entities with IDComponent.
     *
     * @param name entity name
     * @param id entity id
     * @return entity that matches the query or {@link Optional#empty()}
     */
    public Optional<Entity> getEntityByID(String name, int id) {
        for (Entity e : getEntitiesByComponent(IDComponent.class)) {
            IDComponent idComponent = e.getComponent(IDComponent.class);
            if (idComponent.getName().equals(name) && idComponent.getID() == id) {
                return Optional.of(e);
            }
        }

        return Optional.empty();
    }
}

File: fxgl-entity/src/main/kotlin/com/almasb/fxgl/entity/GameWorld.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.entity

import com.almasb.fxgl.core.collection.Array
import com.almasb.fxgl.core.collection.UnorderedArray
import com.almasb.fxgl.core.math.FXGLMath
import com.almasb.fxgl.core.reflect.ReflectionUtils
import com.almasb.fxgl.core.util.Predicate
import com.almasb.fxgl.core.util.tryCatchRoot
import com.almasb.fxgl.entity.component.Component
import com.almasb.fxgl.entity.components.IDComponent
import com.almasb.fxgl.entity.components.IrremovableComponent
import com.almasb.fxgl.entity.components.TimeComponent
import com.almasb.fxgl.entity.level.Level
import com.almasb.sslogger.Logger
import javafx.geometry.Point2D
import javafx.geometry.Rectangle2D
import java.util.*
import kotlin.NoSuchElementException
import kotlin.collections.ArrayList

/**
 * Represents pure logical state of the game.
 * Manages all entities and allows queries.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
class GameWorld {

    companion object {
        private val log = Logger.get("GameWorld")
    }

    private val updateList = Array<Entity>()

    /**
     * List of entities added to the update list in the next tick.
     */
    private val waitingList = UnorderedArray<Entity>()

    /**
     * List of entities in the world.
     *
     * @return direct list of entities in the world (do NOT modify)
     */
    val entities = ArrayList<Entity>()

    /**
     * @return shallow copy of the entities list (new list)
     */
    val entitiesCopy: List<Entity>
        get(): List<Entity> = ArrayList(entities)

    init {
        log.debug("Game world initialized")
    }

    /**
     * The entity will be added to update list in the next tick.
     *
     * @param entity the entity to add to world
     */
    fun addEntity(entity: Entity) {
        require(!entity.isActive){ "Entity is already attached to world" }

        waitingList.add(entity)
        entities.add(entity)

        add(entity)
    }

    fun addEntities(vararg entitiesToAdd: Entity) {
        for (e in entitiesToAdd) {
            addEntity(e)
        }
    }

    private fun add(entity: Entity) {
        entity.init(this)
        notifyEntityAdded(entity)
    }

    /**
     * Entity will be removed from world and parties notified of removal in the same frame.
     * The components and controls of entity will be removed in the next frame to avoid
     * concurrency issues.
     */
    fun removeEntity(entity: Entity) {
        if (!entity.isActive) {
            log.warning("Attempted to remove entity which is not active")
            return
        }

        if (!canRemove(entity))
            return

        require(entity.world === this) { "Attempted to remove entity not attached to this world" }

        entities.remove(entity)

        entity.markForRemoval()
        notifyEntityRemoved(entity)

        // we cannot clean entities here because this may have been called through a control
        // while entity is being updated
        // so, we clean the entity on next frame
    }

    fun removeEntities(vararg entitiesToRemove: Entity) {
        for (e in entitiesToRemove) {
            removeEntity(e)
        }
    }

    fun removeEntities(entitiesToRemove: Collection<Entity>) {
        for (e in entitiesToRemove) {
            removeEntity(e)
        }
    }

    private fun canRemove(entity: Entity): Boolean {
        return !entity.hasComponent(IrremovableComponent::class.java)
    }

    /**
     * Performs a single world update tick.
     *
     * @param tpf time per frame
     */
    fun onUpdate(tpf: Double) {
        updateList.addAll(waitingList)
        waitingList.clear()

        val it = updateList.iterator()
        while (it.hasNext()) {
            val e = it.next()

            if (!e.isActive) {
                // clean entities removed in the last frame
                e.clean()
                it.remove()
            } else {
                val tpfRatio = e.getComponentOptional(TimeComponent::class.java)
                        .map { it.value }
                        .orElse(1.0)

                e.update(tpf * tpfRatio)
            }
        }
    }

    /**
     * Removes all (including with IrremovableComponent) entities.
     * Does NOT clear state listeners.
     * Do NOT call this method manually.
     * It is called automatically by FXGL during initGame().
     */
    fun clear() {
        log.debug("Clearing game world")

        waitingList.clear()

        // we may still have some entities that have been removed but not yet cleaned
        run {
            val it = updateList.iterator()
            while (it.hasNext()) {
                val e = it.next()

                if (!e.isActive) {
                    e.clean()
                }

                it.remove()
            }
        }

        // entities list does not contain "not active" entities, so we do full clean
        // also we copy since during removal notification components may remove other entities
        entitiesCopy.forEach { e ->
            e.markForRemoval()
            notifyEntityRemoved(e)
            e.clean()
        }

        entities.clear()
        entityFactories.clear()
        entitySpawners.clear()
    }

    private val worldListeners = Array<EntityWorldListener>()

    fun addWorldListener(listener: EntityWorldListener) {
        worldListeners.add(listener)
    }

    fun removeWorldListener(listener: EntityWorldListener) {
        worldListeners.removeValueByIdentity(listener)
    }

    private fun notifyEntityAdded(e: Entity) {
        worldListeners.forEach { it.onEntityAdded(e) }
    }

    private fun notifyEntityRemoved(e: Entity) {
        worldListeners.forEach { it.onEntityRemoved(e) }
    }

    /**
     * Set level to given.
     * Resets the world.
     * Adds all level entities to the world.
     *
     * @param level the level
     */
    fun setLevel(level: Level) {
        clearLevel()

        log.debug("Setting level: $level")
        level.entities.forEach { addEntity(it) }
    }

    /**
     * Removes removable entities.
     */
    private fun clearLevel() {
        log.debug("Clearing removable entities")

        waitingList.clear()

        // we may still have some entities that have been removed but not yet cleaned
        // but we do not want to remove Irremovables
        run {
            val it = updateList.iterator()
            while (it.hasNext()) {
                val e = it.next()

                if (canRemove(e)) {
                    if (!e.isActive) {
                        // clean it here because "entities" list does not have "e"
                        e.clean()
                    }

                    it.remove()
                }
            }
        }

        // entities list does not contain "not active" entities, so we do full clean
        // but we do not want to remove Irremovables
        val it = entities.iterator()
        while (it.hasNext()) {
            val e = it.next()

            if (canRemove(e)) {
                e.markForRemoval()
                notifyEntityRemoved(e)
                e.clean()

                it.remove()
            }
        }
    }

    private val entityFactories = hashMapOf<EntityFactory, List<String>>()

    /**
     * Maps entity spawn name to the method that creates the entity.
     */
    private val entitySpawners = hashMapOf<String, EntitySpawner>()

    /**
     * @param entityFactory factory for creating entities
     */
    fun addEntityFactory(entityFactory: EntityFactory) {
        val entityNames = arrayListOf<String>()

        ReflectionUtils.findMethodsMapToFunctions(entityFactory, Spawns::class.java, EntitySpawner::class.java)
                .forEach { annotation, entitySpawner ->

                    val entityAliases = annotation.value.split(",".toRegex())
                    entityAliases.forEach { entityName ->
                        checkDuplicateSpawners(entityFactory, entityName)

                        entitySpawners.put(entityName, entitySpawner)
                        entityNames.add(entityName)
                    }
                }

//        ReflectionUtils.findMethods(entityFactory, Preload::class.java)
//                .forEach { preload, method ->
//
//                    val ann = method.getDeclaredAnnotation(Spawns::class.java)
//                    val entityAliases = ann.value.split(",".toRegex())
//
//                    val numToPreload = preload.value
//
//                    entityPreloader.addForPreloading(entityAliases, numToPreload)
//                }

        entityFactories.put(entityFactory, entityNames)
    }

    private fun checkDuplicateSpawners(entityFactory: EntityFactory, entityName: String) {
        if (entitySpawners.containsKey(entityName)) {

            // find the factory that already has entityName spawner
            val factory = entityFactories.entries.find { entityName in it.value }

            throw IllegalArgumentException("Duplicated @Spawns($entityName) in $entityFactory. Already exists in $factory")
        }
    }

    fun removeEntityFactory(entityFactory: EntityFactory) {
        entityFactories.remove(entityFactory)?.forEach {
            entitySpawners.remove(it)
        }
    }

    /**
     * Creates an entity with given name at 0, 0 using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @return spawned entity
     */
    fun spawn(entityName: String): Entity {
        return spawn(entityName, 0.0, 0.0)
    }

    /**
     * Creates an entity with given name at x, y using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param position spawn location
     * @return spawned entity
     */
    fun spawn(entityName: String, position: Point2D): Entity {
        return spawn(entityName, position.x, position.y)
    }

    /**
     * Creates an entity with given name at x, y using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param x x position
     * @param y y position
     * @return spawned entity
     */
    fun spawn(entityName: String, x: Double, y: Double): Entity {
        return spawn(entityName, SpawnData(x, y))
    }

    /**
     * Creates an entity with given name and data using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param data spawn data, such as x, y and any extra info
     * @return spawned entity
     */
    fun spawn(entityName: String, data: SpawnData): Entity {
        val entity = create(entityName, data)
        addEntity(entity)
        return entity
    }

    /**
     * Creates an entity with given name and data using a previously added entity factory.
     * Does NOT add created entity to the game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param data spawn data, such as x, y and any extra info
     * @return created entity
     */
    fun create(entityName: String, data: SpawnData): Entity {
        check(entityFactories.isNotEmpty()) { "No EntityFactory was added! Call gameWorld.addEntityFactory()" }

        val spawner = entitySpawners.get(entityName)
                ?: throw IllegalArgumentException("No EntityFactory has a method annotated @Spawns($entityName)")

        if (!data.hasKey("type")) {
            data.put("type", entityName)
        }

//        if (entityPreloader.isPreloadingEnabled(entityName)) {
//            return entityPreloader.obtain(entityName, data)
//        }

        return tryCatchRoot { spawner.apply(data) }
    }

    /* QUERIES */

    fun getSingleton(type: Enum<*>): Entity {
        return getSingleton(Predicate { it.isType(type) })
    }

    fun getSingleton(predicate: Predicate<Entity>): Entity {
        return entities.find { predicate.test(it) } ?: throw NoSuchElementException("No entity found satisfying the predicate")
    }

    /**
     * Useful for singleton type entities, e.g. Player.
     *
     * @return first occurrence matching given type
     */
    fun getSingletonOptional(type: Enum<*>): Optional<Entity> {
        return getSingletonOptional(Predicate { it.isType(type) })
    }

    /**
     * @return first occurrence matching given predicate
     */
    fun getSingletonOptional(predicate: Predicate<Entity>): Optional<Entity> {
        return Optional.ofNullable(entities.find { predicate.test(it) })
    }

    /**
     * @param type entity type
     * @return a random entity with given type
     */
    fun getRandom(type: Enum<*>): Optional<Entity> {
        return FXGLMath.random(getEntitiesByType(type))
    }

    fun getRandom(predicate: Predicate<Entity>): Optional<Entity> {
        return FXGLMath.random(getEntitiesFiltered(predicate))
    }

    /**
     * @param type component type
     * @return array of entities that have given component
     */
    fun getEntitiesByComponent(type: Class<out Component>): List<Entity> {
        return entities.filter { it.hasComponent(type) }
    }

    /**
     * Returns a list of entities which are filtered by
     * given predicate.
     *
     * @param predicate filter
     * @return new list containing entities that satisfy query filters
     */
    fun getEntitiesFiltered(predicate: Predicate<Entity>): List<Entity> {
        return entities.filter { predicate.test(it) }
    }

    /**
     * If called with no arguments, all entities are returned.
     *
     * @param types entity types
     * @return new list containing entities that satisfy query filters
     */
    fun getEntitiesByType(vararg types: Enum<*>): List<Entity> {
        if (types.isEmpty())
            return entitiesCopy

        return entities.filter { isOneOfTypes(it, *types) }
    }

    private fun isOneOfTypes(entity: Entity, vararg types: Enum<*>): Boolean {
        return types.any { entity.isType(it) }
    }

    /**
     * Returns a list of entities
     * which are partially or entirely
     * in the specified rectangular selection.
     * Warning: object allocation.
     *
     * @param selection Rectangle2D that describes the selection box
     * @return new list containing entities that satisfy query filters
     */
    fun getEntitiesInRange(selection: Rectangle2D): List<Entity> {
        return entities.filter { it.boundingBoxComponent.isWithin(selection) }
    }

    /**
     * Returns a list of entities
     * which colliding with given entity.
     *
     * Note: CollidableComponent is not considered.
     *
     * @param entity the entity
     * @return new list containing entities that satisfy query filters
     */
    fun getCollidingEntities(entity: Entity): List<Entity> {
        return entities.filter { it.isColliding(entity) && it !== entity }
    }

    /**
     * Returns a list of entities at given position.
     * The position x and y must equal to entity's position x and y.
     *
     * @param position point in the world
     * @return entities at given point
     */
    fun getEntitiesAt(position: Point2D): List<Entity> {
        return entities.filter { it.position == position }
    }

    /**
     * Returns the closest entity to the given entity with given
     * filter. The given
     * entity itself is never returned.
     *
     * If there no entities satisfying the requirement, [Optional.empty]
     * is returned.
     *
     * @param entity selected entity
     * @param filter requirements
     * @return closest entity to selected entity with type
     */
    fun getClosestEntity(entity: Entity, filter: Predicate<Entity>): Optional<Entity> {
        return Optional.ofNullable(
                entities.filter { filter.test(it) && it !== entity }
                        .sortedBy { entity.distance(it) }
                        .firstOrNull()
        )
    }

    /**
     * Returns an entity whose IDComponent matches given name and id.
     *
     *
     * Returns [Optional.empty] if no entity was found with such combination.
     * This query only works on entities with IDComponent.
     *
     * @param name entity name
     * @param id entity id
     * @return entity that matches the query or [Optional.empty]
     */
    fun getEntityByID(name: String, id: Int): Optional<Entity> {
        return Optional.ofNullable(
                getEntitiesByComponent(IDComponent::class.java).find {
                    val idComponent = it.getComponent(IDComponent::class.java)

                    return@find idComponent.name == name && idComponent.id == id
                }
        )
    }
}